src/storage/src/source/source_reader_pipeline.rs

```diff
@@ -1277,287 +804,149 @@ where
-                        Err(ReclockError::BeyondUpper(_) | ReclockError::Uninitialized) => {
-                            trace!(
-                                "reclock({id}) {worker_id}/{worker_count}: \
-                                cannot yet reclock batch with \
-                                batch_upper: {:?} \
-                                batch_lower: {:?} \
-                                reclock.source_frontier: {:?}",
-                                untimestamped_batch.batch_upper,
-                                untimestamped_batch.batch_lower,
-                                timestamper.source_upper()
-                            );
-                            // We keep batches in the order they arrive from the
-                            // source. And we assume that the source frontier never
-                            // regressses. So we can break now.
-                            break;
-                        }
-                        Err(ReclockError::NotBeyondSince(ts)) => {
-                            panic!("failed to reclock time {:?}", ts)
-                        }
-                    };
+                    // If the remap frontier advanced it's time to carve out a batch that includes
+                    // all updates not beyond the upper
+                    Event::Progress(remap_upper) => {
+                        let remap_trace_batch = ReclockBatch {
+                            updates: remap_updates_stash
+                                .drain_filter_swapping(|(_, ts, _)| !remap_upper.less_equal(ts))
+                                .collect(),
+                            upper: remap_upper.to_owned(),
+                        };
+                        trace!("timely-{worker_id} reclock({id}) received remap batch: updates={:?} upper={}", &remap_trace_batch.updates, remap_upper.pretty());
+                        timestamper.push_trace_batch(remap_trace_batch);
+                        work_to_do.notify_one();
+                    }
+                },
+                Some(event) = source_rx.recv() => match event {
+                    EventCore::Progress(changes) => {
+                        source_upper.update_iter(changes);
+                        trace!("timely-{worker_id} reclock({id}) received source progress: source_upper={}", source_upper.pretty());
+                        work_to_do.notify_one();
+                    }
+                    EventCore::Messages(_time, batch) => {
+                        untimestamped_batches.push(batch)
+                    }
+                },
+                _ = work_to_do.notified(), if timestamper.initialized() => {
+                    // The frontier at which we can retire work is the minimum of the timestamper
+                    // upper and the source data upper
+                    let mut ready_upper  = timestamper.source_upper();
+                    ready_upper.extend(source_upper.frontier().iter().cloned());
```

Comment:
could we do an explicit `meet`? Or is it too onerous to have a `Lattice` bound on the source times here?

Reply:
I wouldn't put a `Lattice` bound just for meet, since this is supported out of the box. But also `Partitioned` timestamps are not `Lattice` which would be an additional complication